synapse/storage/databases/state/store.py

```diff
@@ -106,6 +119,12 @@ def __init__(
             500000,
         )
 
+        # Current ongoing get_state_for_groups in-flight requests
+        # {group ID -> {StateFilter -> ObservableDeferred}}
+        self._state_group_inflight_requests: Dict[
+            int, Dict[StateFilter, ObservableDeferred[StateMap[str]]]
```

Comment:
Random driveby comment: Consider typing this with `AbstractObservableDeferred` instead:

```suggestion
        self._state_group_inflight_requests: Dict[
            int, Dict[StateFilter, AbstractObservableDeferred[StateMap[str]]]
```

Once you've created the `ObservableDeferred`, the only thing you should be doing with it is `.observe()`ing it, which you can enforce via `AbstractObservableDeferred`.

(somewhat related to my campaign on https://github.com/matrix-org/synapse/issues/11390)